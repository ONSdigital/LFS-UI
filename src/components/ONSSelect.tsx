import React, { Component, ChangeEvent } from 'react';

interface Props{
  label?: string,
  id?: string,
  onChange? : (e: ChangeEvent<HTMLSelectElement>) => {},
  value: string,
  options: Option[]
}

interface Option{
  label:string,
  value?:string
}

interface State{
  value: string
}

export class ONSSelect extends Component <Props, State>{
  value: string = this.props.value !== undefined ? this.props.value : "";
  constructor(props : Props) {
    super(props);
    this.state = {value: props.value !== undefined ? this.props.value : ""};
  }

  handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
    if(this.props.onChange !== undefined){
      this.props.onChange(e);
    }
    this.value = e.target.value;
    this.setState({value: e.target.value})
  }

    render() {
      return (
        <p className="field">
          {this.props.label !== undefined &&
            <label className="label">{this.props.label}</label>
          }
          <select id="select" name="select" defaultValue="" className="input input--select" onChange={(e) => this.handleChange(e)}>
            <option value="" disabled>
              Select an option
            </option>
            {this.props.options.map((option, index) =>
              <option value={option.value} key={index}>
                {option.label}    
              </option>
            )}
          </select>
        </p>
      );
  }
}
