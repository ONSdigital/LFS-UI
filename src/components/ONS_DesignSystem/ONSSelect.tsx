import React, {ChangeEvent, Component} from 'react';

interface Props{
  label?: string,
  id?: string,
  onChange? : (e: ChangeEvent<HTMLSelectElement>, ...args: any[]) => void,
  value: string,
  options: Option[]
}

interface Option{
  label:string,
  value?:string
  id?: string
}

interface State{
  value: string
}

export class ONSSelect extends Component <Props, State>{
  value: string = this.props.value !== undefined ? this.props.value : "";
  constructor(props : Props) {
    super(props);
    this.state = {value: props.value !== undefined ? this.props.value : ""};
  }

  handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
    if(this.props.onChange !== undefined){
      this.props.onChange(e);
    }
    this.value = e.target.value;
    this.setState({value: e.target.value})
  };

    render() {
      return (
        <p className="field">
          {this.props.label !== undefined &&
            <label className="label">{this.props.label}</label>
          }
          <select id="select" name="select" defaultValue="" className="input input--select" onChange={(e) => this.handleChange(e)}>
            <option value="" disabled>
              Select an option
            </option>
            {this.props.options.map((option, index) =>
              <option value={option.value} key={index} id={option.id}>
                {option.label}    
              </option>
            )}
          </select>
        </p>
      );
  }
}
