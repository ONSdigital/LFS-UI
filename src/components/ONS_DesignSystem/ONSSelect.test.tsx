import React from "react";
import Enzyme, {mount, shallow} from "enzyme";
import {ONSSelect} from "./ONSSelect";
import Adapter from "enzyme-adapter-react-16";

describe("ONS Select Test", () => {
    Enzyme.configure({ adapter: new Adapter() });

    const Selection = [
        {"label": "1", "value": "1"},
        {"label": "2", "value": "2"},
        {"label": "3", "value": "3"}
    ]

    const Props = {
        label: "Select From",
        value: "select value",
        options: Selection
    }

    const changeProps = {
        label: "Select From",
        options: Selection,
        onChange: jest.fn(),
    }

    const undefinedChangeProps = {
        label: "Select From",
        value: "select value",
        options: Selection,
        onChange: undefined,
    }

    function wrapper(render: any, props: any) {
        return render(
            <ONSSelect
                label={props.label}
                onChange={props.onChange}
                value={props.value}
                options={props.options}>
            </ONSSelect>
            
        )
    }

    it("matches Snapshot", () => {
        expect(wrapper(shallow, Props)).toMatchSnapshot()
    });

    it("should render correctly", () => expect(wrapper(shallow, Props).exists()).toEqual(true));
    
    it("should render with the correct label", () => {
        expect(wrapper(mount, Props).find("ONSSelect").getElement().props.label).toEqual(Props.label);
    })
    
    it('simulates change events', () => {
        //defined
        wrapper(mount, changeProps).find("select").simulate('change');
        expect(changeProps.onChange).toHaveBeenCalled()
        
        //undefined
        wrapper(mount, undefinedChangeProps).find("select").simulate('change');
        expect(undefinedChangeProps.onChange).toBeUndefined()
    })
})