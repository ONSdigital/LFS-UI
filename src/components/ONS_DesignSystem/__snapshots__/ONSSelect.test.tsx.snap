// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ONS Select Test matches Snapshot 1`] = `ShallowWrapper {}`;
